### ISSUE TITLE
Lack of ability to pitch bend & pan inside Piano-Roll for ZASFX, LB302 & OpulenZ: idea for workaround

### ISSUE THREAD

**Issue by EtornaZ** [LMMS/lmms#3955]
As we know, ZASFX, LB302 and OpulenZ don't have the ability to pan and pitch bend inside the Piano Roll. I was thinking about this and I have an idea as a workaround. 

What if the pitch knob (or even an invisible copy) for those synthesizers had its value connected to the pitch bending automation editor inside the Piano Roll, likewise with the panning knob's value on the note-panning sliders (or whatever they're called) in the Piano Roll? 
Maybe the volume and panning sliders in the Piano Roll could even be replaced with an automation editor, like the one for pitch bending. This would make it easier for individual notes to have a rise and fall of panning and volume.

I have zero idea on how the coding for LMMS works, but I am willing to learn if needed; I've played around with XML so I get the structure of coding.

**Comment by musikBear** [comment]
Njea I actually think lmms should rethink current pitch-bend edits. It would be so much simpler just to drop n objects on piano-roll for a selected note. Currently it is buggy to edit the pitch. A transparent editor is almost a must, so bottom line : Rethink UI for pitch-bend

**Comment by tresf** [comment]
> Rethink UI for pitch-bend

This is out of scope.  What the OP is asking is to leverage a piano-style pitch bend on instruments that are piano-roll pitchable.  The UX implementation of the current piano pitch-bend can receive improvements regardless of this feature.

That said, I agree that the current implementation is very difficult to use in a MDI-style layered windowing environment.

**Comment by EtornaZ** [comment]
I agree with musikBear. Dragging values, specifically the panning, pitch and volume ones, onto individual notes in the Piano Roll and then having an automation editor open up would be a better solution than my idea.

**Comment by tresf** [comment]
@EtornaZ I believe the problem with the piano-roll pitchbends is that it's all all-or-nothing problem.  We could leverage pitchbends in piano roll, but they wouldn't be per-note, they'd only be percieved visually as per-note.  If multiple notes were to be pitchbent at the same time, they'd be in a conflict with one another.

I think simply adding a warning to the user (perhaps something non-interupting) that the instrument does not support per-note pitchbends and allowing the automation editor to open is fine from a UX perspective.

Another slight is that if the same instrument were used in multiple areas at once (e.g. BB Editor) one pitchbend would affect all instances of that instrument.  Whatever though.  It would be nice to have a common interface and deal with the instrument limitations by making them obvious in the UI/UX.

**Comment by EtornaZ** [comment]
@tresf Alright, that's cool. :)
So how would you go about this? Just fix the UX so that the aforementioned synthesizers will pitch bend in the piano roll?

**Comment by tresf** [comment]
@EtornaZ that would be my vote, yes, but we would need a clear visual indicator to that instrument at that exact time that the pitch is being messed with so it's easy to find in the event of simultaneous instrument re-use such as stacked BB Tracks.

**Comment by tresf** [comment]
@EtornaZ that would be my vote, yes, but we would need a clear visual indicator to that instrument at that exact time that the pitch is being messed with so it's easy to find in the event of simultaneous instrument re-use such as stacked BB Tracks.

**Comment by EtornaZ** [comment]
@tresf Maybe the notes in the Song-Editor could have the pitch bent lines shown? Same with B&Bs, they could be made transparent in the Song-Editor if they have notes in them; that way, people could see the notes and pitch bends in them. Yes, I know if a B&B editor has more than one instrument/synthesizer with notes that could be visually messy, but people can stretch the B&B Editor in the Song-Editor and see all of the notes.

**Comment by EtornaZ** [comment]
@tresf Maybe the notes in the Song-Editor could have the pitch bent lines shown? Same with B&Bs, they could be made transparent in the Song-Editor if they have notes in them; that way, people could see the notes and pitch bends in them. Yes, I know if a B&B editor has more than one instrument/synthesizer with notes that could be visually messy, but people can stretch the B&B Editor in the Song-Editor and see all of the notes.

**Comment by gi0e5b06** [comment]
Every controller should be both setable per note and per automation track. If there is more than one setting, all settings should be combined.

**Comment by JohannesLorenz** [comment]
I think this should be implemented by letting single notes execute the global automation. This will be mostly the same as per-note-pitch in case only one note as a time is being played.

* If there is currently no automation, keep the pitch as it was previously.
* If a new note with automation is being added, and no other note with automation is playing, its automation is being used.
* If a new note with automation is being added and another note with automation is playing, the new note is being added to a stack, giving it precedence over previous notes.

There shall be no quick jumps, so if a new note comes in with a different pitch, the pitch must be interpolated.

Can we implement it like that?

**Comment by d3im** [comment]
Hi, just looked why note panning doesn't work with Sf2 - FluidR3. Any chance to resolve this soon? Thanks.
